### DISCUSSION TITLE
Call fetch in server component when going back

### DISCUSSION THREAD

**Discussion by jim-king-2000** [vercel/next.js#67937] [Help]
### Summary

Suppose we have a fetch call in the server component:
```jsx
export async function Page() {
  const data = await fetch('https://xxxx.com/');
  return (
    <>
      <DataPanel data={data} />
      <Link href='/next' />
    </>
  );
}
```

When we click the link and navigate to the '/next' page, and then go back by calling `router.back()` (or click the back button in the browser), the fetch is **NOT** get invoked and the data is stale.

Is it possible to call fetch every time when navigating back to the page?

**Comment by icyJoseph** [comment]
Could you try with: https://nextjs.org/docs/app/api-reference/next-config-js/staleTimes

Set them both to zero, and report back what happens.

**Comment by jim-king-2000** [comment]
Set them both to 0, go back still can't trigger server fetch. Even worse, Link does not trigger server fetch either.

**Reply by icyJoseph** [reply]
Ok, the server fetch is possibly cached. Do the request like: `fetch('https://...', { cache: 'no-store' })`

**Comment by jim-king-2000** [comment]
The fetch is encapsulated in an SDK. If we have to add `{ cache: 'no-store' }` here, we'd like to add it to every fetch of every SDK. That would be a huge impact which we'd better to avoid.
Since `staleTimes.dynamic: 0` could work for `<Link>`, is it possible that Next.js team let it also work for `router.back()`?

**Comment by jim-king-2000** [comment]
I found `staleTimes.dynamic: 0` is unstable. When clicking the link, some time it refreshes the data, some time not.

**Reply by icyJoseph** [reply]
Funny, because IIRC, in Next 15 they made that the default.

> In Next.js 14.2.0, we introduced an experimental [staleTimes](https://rc.nextjs.org/docs/app/api-reference/next-config-js/staleTimes) flag to allow custom configuration of the [Router Cache](https://rc.nextjs.org/docs/app/building-your-application/caching#router-cache).
> 
> In Next.js 15, this flag still remains accessible, but we are changing the default behavior to have a staleTime of 0 for Page segments. This means that as you navigate around your app, the client will always reflect the latest data from the Page component(s) that become active as part of the navigation. However, there are still important behaviors that remain unchanged:
> 
> Shared layout data won't be refetched from the server to continue to support [partial rendering](https://rc.nextjs.org/docs/app/building-your-application/routing/linking-and-navigating#4-partial-rendering).
> Back/forward navigation will still restore from cache to ensure the browser can restore scroll position.
> [Loading.js](https://rc.nextjs.org/docs/app/api-reference/file-conventions/loading) will remain cached for 5 minutes (or the value of the staleTimes.static configuration).

- https://nextjs.org/blog/next-15-rc

That's probably not the real issue though.

> The fetch is encapsulated in an SDK. If we have to add { cache: 'no-store' } here, we'd like to add it to every fetch of every SDK. That would be a huge impact which we'd better to avoid.

Could you do the page level configuration: 

- https://nextjs.org/docs/app/api-reference/file-conventions/route-segment-config#fetchcache

**Reply by jim-king-2000** [reply]
Finally I choose `fetch('https://...', { cache: 'no-store' })`. It works well and is very stable. I spent too much time debugging the cache issue. I almost disabled all of the caches (export const dynamic = 'force-dynamic'; staleTimes.dynamic: 0; and { cache: 'no-store' }) Next.js v14 introduces. And the performance of the web page is not impacted at all.

**Reply by jim-king-2000** [reply]
I find `fetch('https://...', { cache: 'no-store' })` is still unstable. It makes me crazy...

**Reply by jim-king-2000** [reply]
Added the following lines to the layout, they do not work:
```javascript
export const dynamic = 'force-dynamic';
export const revalidate = 0;
export const fetchCache = 'default-no-store';
```

**Reply by jim-king-2000** [reply]
`router.push()` and `<a>` always works. `<Link>` and `router.back()` works intermittently.

**Reply by jim-king-2000** [reply]
By the way, the `<Link>` and `router` is from `next-intl/navigation`.

**Reply by icyJoseph** [reply]
🤔 I am not well versed in the implementation details of, https://web.dev/articles/bfcache, but could you try to read about it, and debugging techniques, to see if that's what's happening?

**Reply by jim-king-2000** [reply]
I add the following code snippet to my project. But nothing is printed. I don't think it is relevant to bfcache.
```javascript
window.addEventListener('pageshow', (event) => {
	if (event.persisted) {
		console.log('This page was restored from the bfcache.');
	} else {
		console.log('This page was loaded normally.');
	}
});
```

**Reply by jim-king-2000** [reply]
Even `router.push()` from `next/navigation` would **NOT** 100% purge the cache.

**Reply by jim-king-2000** [reply]
`<a>` can 100% purge server fetch cache.

**Reply by icyJoseph** [reply]
but when you do the navigations that fail, do you see `_rsc` requests? they usually have an `_rsc` query attached to them, and represent the next tree to render.

**Reply by jim-king-2000** [reply]
If I click the "back" button of the Chrome, there is no `_rsc` requests. But if I click the `<Link>`, there is a `_rsc` request issued.

**Comment by jim-king-2000** [comment]
Any update or workaround?
